File: quick_sort.py
def partition(arr, startIndex, stopIndex):
    pivot = arr[stopIndex]
    partitionIndex = startIndex

    for index in range(startIndex, stopIndex):
        if arr[index] <= pivot:
            arr[partitionIndex], arr[index] = arr[index], arr[partitionIndex]
            partitionIndex = partitionIndex+1

    arr[partitionIndex], arr[stopIndex] = arr[stopIndex], arr[partitionIndex]
    return partitionIndex


def quickSort(arr, startIndex, stopIndex):
    if startIndex < stopIndex:
        pivot = partition(arr, startIndex, stopIndex)
        quickSort(arr, startIndex, pivot-1)
        quickSort(arr, pivot+1, stopIndex)


test_arr = [10, 7, 8, 9, 1, 5]
quickSort(test_arr, 0, len(test_arr)-1)
print(test_arr)

File: sorting-algo/quickSort.js
function partition(arr, start, end) {
    let pivot = arr[end];
    let pivotIndex = start - 1;
    for (let i = start; i <= end; i++) {
        if (arr[i] < pivot) {
            let temp = arr[i];
            arr[i] = arr[pivotIndex];
            arr[pivotIndex] = temp;
            pivotIndex++;
        }
    }
    let temp = arr[end];
    arr[end] = arr[pivotIndex + 1];
    arr[pivotIndex + 1] = temp;
    return pivotIndex + 1;
}

function quickSort(arr, start, end) {
    if (start < end) {
        let partitionIndex = partition(arr, start, end);
        quickSort(arr, start, partitionIndex - 1);
        quickSort(arr, partitionIndex, end);
    }
}


let arr = [2, 4, 5, 60, 70, 7, 32, 36, 1, 8, 9];
quickSort(arr)

console.log(arr);
